### PR TITLE
ipc: remove waiti running in IRQ level higher than passive level

### PR DESCRIPTION
**Pull request by RanderWang**
waiti would cause FW panic if it is called in irq level
higher than passive level

Signed-off-by: Rander Wang <rander.wang@linux.intel.com>

### CODE REVIEW

File: src/ipc/ipc.c [hunk 1]
```diff
@@ -501,8 +502,15 @@ int ipc_get_page_descriptors(struct dma *dmac, uint8_t *page_table,
...
-	complete.timeout = PLATFORM_HOST_DMA_TIMEOUT;
-	ret = wait_for_completion_timeout(&complete);
+	while (!wait_is_completed(&complete)) {
+		idelay(PLATFORM_DMA_TIMEOUT);
+
+		if (try++ > 10)
+			break;
+	}
+
+	if (!wait_is_completed(&complete))
```
Comment (lgirdwood) [inline src/ipc/ipc.c]: trace_error needed here.

File: src/ipc/ipc.c [hunk 2]
```diff
@@ -501,8 +502,15 @@ int ipc_get_page_descriptors(struct dma *dmac, uint8_t *page_table,
 	}
 
 	/* wait for DMA to complete */
-	complete.timeout = PLATFORM_HOST_DMA_TIMEOUT;
-	ret = wait_for_completion_timeout(&complete);
+	while (!wait_is_completed(&complete)) {
+		idelay(PLATFORM_DMA_TIMEOUT);
```
[the remark this answers is not in the document]
Comment (lgirdwood) [reply src/ipc/ipc.c]: PLATFORM_HOST_DMA_TRIES

[the remark this answers is not in the document]
Comment (RanderWang) [reply src/ipc/ipc.c]: ok, I will refine my patch

File: src/ipc/ipc.c [hunk 3]
```diff
@@ -501,8 +502,15 @@ int ipc_get_page_descriptors(struct dma *dmac, uint8_t *page_table,
...
+
+		if (try++ > 10)
```
[the remark this answers is not in the document]
Comment (RanderWang) [reply src/ipc/ipc.c]: ok, I will refine my patch

File: src/ipc/ipc.c (hunk 1 above)
Comment (RanderWang) [reply src/ipc/ipc.c]: ok, I will refine my patch

File: src/include/sof/wait.h [hunk 4]
```diff
@@ -145,6 +146,25 @@ static inline int wait_for_completion_timeout(completion_t *comp)
 	}
 }
 
+static inline int wait_for_completion_idelay(int (*condition)(void *data), void *data, int clk)
```
Comment (lgirdwood) [inline src/include/sof/wait.h]: int clk should be named delay_ms or delay_us depending on time measurement.

File: src/include/sof/wait.h [hunk 5]
```diff
@@ -145,6 +146,25 @@ static inline int wait_for_completion_timeout(completion_t *comp)
...
+{
+	int try = PLATFORM_DEFAULT_TRIES;
+
+	while (!condition(data)) {
+		idelay(clk);
```
Comment (lgirdwood) [inline src/include/sof/wait.h]: this must be divided by tries. otherwise we will wait for multiples of the timeout and not the timeout.

**Comment by RanderWang** [comment]
update my patch, thanks for review

File: src/include/sof/wait.h (hunk 5 above)
Comment (RanderWang) [reply src/include/sof/wait.h]: yes, I will refine my patch

File: src/include/sof/wait.h (hunk 4 above)
Comment (RanderWang) [reply src/include/sof/wait.h]: yes, I will refine my patch

File: src/include/sof/wait.h [hunk 6]
```diff
@@ -157,4 +160,17 @@ static inline void wait_delay(uint64_t number_of_clks)
 		idelay(PLATFORM_DEFAULT_DELAY);
 }
 
+static inline int wait_for_completion_udelay(int (*condition)(void *data), void *data, int us)
```
Comment (lgirdwood) [inline src/include/sof/wait.h]: why not pass the completion_t here instead of a callback, makes it simpler

Comment (RanderWang) [reply src/include/sof/wait.h]: because in some case, there is no completion, just like waiting for a register idle

Comment (RanderWang) [reply src/include/sof/wait.h]: please refer to my second commit in this PR

Comment (lgirdwood) [reply src/include/sof/wait.h]: waiting for a register is called polling, and this is probably better named poll_for_condition_timeout()

Comment (RanderWang) [reply src/include/sof/wait.h]: I just want to make a generic solution.  Ok, I will refine my patch

**Comment by RanderWang** [comment]
update my patch, thanks for review

File: src/include/sof/wait.h [hunk 7]
```diff
@@ -157,4 +160,17 @@ static inline void wait_delay(uint64_t number_of_clks)
 		idelay(PLATFORM_DEFAULT_DELAY);
 }
 
+static inline int wait_for_completion_udelay(completion_t *comp, int us)
+{
+	int tick = clock_us_to_ticks(CLK_CPU, us);
+
+	wait_delay(tick);
```
Comment (lgirdwood) [inline src/include/sof/wait.h]: Why did you remove the tries ? and while loop ? This will now block for timeout and be horribly inefficient. We need to try an return as soon as condition completes.

**Comment by RanderWang** [comment]
update my patch, thanks for review

File: src/include/sof/wait.h [hunk 8]
```diff
@@ -157,4 +162,27 @@ static inline void wait_delay(uint64_t number_of_clks)
 		idelay(PLATFORM_DEFAULT_DELAY);
 }
 
+static inline int wait_for_completion_udelay(completion_t *comp, int us)
+{
+	int tick = clock_us_to_ticks(CLK_CPU, us);
+	int try = DEFAULT_TRY_TIMES;
+	int delta = tick / DEFAULT_TRY_TIMES;
+
+	while (1) {
```
Comment (lgirdwood) [inline src/include/sof/wait.h]: while (tries--)

File: src/include/sof/wait.h [hunk 9]
```diff
@@ -157,4 +162,27 @@ static inline void wait_delay(uint64_t number_of_clks)
 		idelay(PLATFORM_DEFAULT_DELAY);
 }
 
+static inline int wait_for_completion_udelay(completion_t *comp, int us)
+{
+	int tick = clock_us_to_ticks(CLK_CPU, us);
+	int try = DEFAULT_TRY_TIMES;
+	int delta = tick / DEFAULT_TRY_TIMES;
```
Comment (lgirdwood) [inline src/include/sof/wait.h]: how do we guarantee delta is not 0 ?

File: src/include/sof/wait.h [hunk 10]
```diff
@@ -157,4 +162,27 @@ static inline void wait_delay(uint64_t number_of_clks)
 		idelay(PLATFORM_DEFAULT_DELAY);
 }
 
+static inline int wait_for_completion_udelay(completion_t *comp, int us)
```
Comment (lgirdwood) [inline src/include/sof/wait.h]: poll_for_completion_delay

File: src/include/sof/wait.h [hunk 11]
```diff
@@ -157,4 +162,27 @@ static inline void wait_delay(uint64_t number_of_clks)
...
+{
+	int tick = clock_us_to_ticks(CLK_CPU, us);
+	int try = DEFAULT_TRY_TIMES;
+	int delta = tick / DEFAULT_TRY_TIMES;
+
+	while (1) {
+		wait_delay(delta);
+		if (wait_is_completed(comp))
+			break;
```
Comment (lgirdwood) [inline src/include/sof/wait.h]: return 0;

File: src/include/sof/wait.h (hunk 9 above)
Comment (RanderWang) [reply src/include/sof/wait.h]: en, a very tricky corner case, I will refine it.

File: src/include/sof/wait.h (hunk 8 above)
Comment (RanderWang) [reply src/include/sof/wait.h]: yes, it is good for me

File: src/include/sof/wait.h (hunk 11 above)
Comment (RanderWang) [reply src/include/sof/wait.h]: get it

File: src/include/sof/wait.h (hunk 10 above)
Comment (RanderWang) [reply src/include/sof/wait.h]: get it

**Comment by RanderWang** [comment]
update my patch, thanks for review

File: src/include/sof/wait.h [hunk 12]
```diff
@@ -157,4 +162,29 @@ static inline void wait_delay(uint64_t number_of_clks)
...
+	int delta = tick / tries;
+
+	if (!delta) {
+		delta = us;
+		tries = 1;
+	}
+
+	while (tries--) {
+		wait_delay(delta);
+		if (wait_is_completed(comp))
```
Comment (lgirdwood) [inline src/include/sof/wait.h]: We should not delay if completion is initially true. This is better.
```C
while (!wait_is_completed(comp)) {
	
	if (!tries--) {
		trace_error();
		return error
	}
		
	delay();
}

return 0;
```

File: src/include/sof/wait.h [hunk 13]
```diff
@@ -157,4 +162,27 @@ static inline void wait_delay(uint64_t number_of_clks)
 		idelay(PLATFORM_DEFAULT_DELAY);
 }
 
+static inline int poll_for_completion_delay(completion_t *comp, uint64_t us)
+{
+	uint64_t tick = clock_us_to_ticks(CLK_CPU, us);
```
Comment (tlauda) [inline src/include/sof/wait.h]: @RanderWang @lgirdwood I think this function won't work correctly, because on cAVS platforms wait_delay function is based on XTAL (e.g. 19.2 MHZ for APL) and CLK_CPU has entirely different freq.

Comment (lgirdwood) [reply src/include/sof/wait.h]: @RanderWang can you fix, use the PLATFORM_CLK thet we use fr work/scheduling.
